Makefile

```diff
@@ -126,80 +114,219 @@ acncli: acncli-binary acncli-archive
 
 # Build the Azure CNI IPAMV6 binary.
-.PHONY: azure-vnet-ipamv6-binary
 azure-vnet-ipamv6-binary:
 	cd $(CNI_IPAMV6_DIR) && CGO_ENABLED=0 go build -v -o $(CNI_BUILD_DIR)/azure-vnet-ipamv6$(EXE_EXT) -ldflags "-X main.version=$(VERSION)" -gcflags="-dwarflocationlists=true"
 
 # Build the Azure CNI telemetry binary.
-.PHONY: azure-vnet-telemetry-binary
 azure-vnet-telemetry-binary:
 	cd $(CNI_TELEMETRY_DIR) && CGO_ENABLED=0 go build -v -o $(CNI_BUILD_DIR)/azure-vnet-telemetry$(EXE_EXT) -ldflags "-X main.version=$(VERSION) -X $(CNI_AI_PATH)=$(CNI_AI_ID)" -gcflags="-dwarflocationlists=true"
 
 # Build the Azure CLI network binary.
-.PHONY: acncli-binary
 acncli-binary:
 	cd $(ACNCLI_DIR) && CGO_ENABLED=0 go build -v -o $(ACNCLI_BUILD_DIR)/acn$(EXE_EXT) -ldflags "-X main.version=$(VERSION)" -gcflags="-dwarflocationlists=true"
 
 # Build the Azure CNS binary.
-.PHONY: azure-cns-binary
 azure-cns-binary:
 	cd $(CNS_DIR) && CGO_ENABLED=0 go build -v -o $(CNS_BUILD_DIR)/azure-cns$(EXE_EXT) -ldflags "-X main.version=$(VERSION) -X $(CNS_AI_PATH)=$(CNS_AI_ID)" -gcflags="-dwarflocationlists=true"
 
 # Build the Azure NPM binary.
-.PHONY: azure-npm-binary
 azure-npm-binary:
 	cd $(CNI_TELEMETRY_DIR) && CGO_ENABLED=0 go build -v -o $(NPM_BUILD_DIR)/azure-vnet-telemetry$(EXE_EXT) -ldflags "-X main.version=$(VERSION)" -gcflags="-dwarflocationlists=true"
 	cd $(NPM_DIR) && CGO_ENABLED=0 go build -v -o $(NPM_BUILD_DIR)/azure-npm$(EXE_EXT) -ldflags "-X main.version=$(VERSION) -X $(NPM_AI_PATH)=$(NPM_AI_ID)" -gcflags="-dwarflocationlists=true"
 
 
-########################### Container Images ########################
+##@ Containers
+
+CNI_IMAGE = azure-cni-manager
+CNS_IMAGE = azure-cns
+NPM_IMAGE = azure-npm
 
-# Build the tools images
-.PHONY: tools-images
-tools-images:
-	$(MKDIR) $(IMAGE_DIR)
-	docker build --no-cache -f ./tools/acncli/Dockerfile --build-arg VERSION=$(VERSION) -t $(AZURE_CNI_IMAGE):$(VERSION) .
-	docker save $(AZURE_CNI_IMAGE):$(VERSION) | gzip -c > $(IMAGE_DIR)/$(CNI_IMAGE_ARCHIVE_NAME)
+TAG               ?= $(VERSION)
+IMAGE_REGISTRY    ?= acnpublic.azurecr.io
+PLATFORM          ?= $(GOOS)/$(GOARCH)
+CONTAINER_BUILDER  = buildah
+
+# prefer buildah, if available, but fall back to docker if that binary is not in the path.
+ifeq (, $(shell which $(CONTAINER_BUILDER)))
+CONTAINER_BUILDER = docker
+endif
+
+## This section is for building individual container images.
+
+container-platform-tag: # util target to print the container tag
+	@echo $(subst /,-,$(PLATFORM))-$(TAG)
+
+containerize-buildah: # util target to build container images using buildah. do not invoke directly.
+	buildah build \
+		--jobs 16 \
+		--platform $(PLATFORM) \
+		-f $(DOCKERFILE) \
+		--build-arg VERSION=$(VERSION) $(EXTRA_BUILD_ARGS) \
+		-t $(REGISTRY)/$(IMAGE):$(TAG) \
+		.
+
+containerize-docker: # util target to build container images using docker buildx. do not invoke directly.
+	docker buildx build \
```

Comment:
I think to truly allow multi-arch images, we should have a `--push` flag option here, otherwise I think it builds the docker images as whatever the local machine architecture is.

Reply:
the `--platform` flag lets us specify the architecture to build, `--push` tells buildx to go ahead and push it to the container registry when it's done building.

Reply:
Can we allow for `--push` to be passed in?

Reply:
no, push is a separate explicit target. if the build step pushes, it will push whatever os/arch it just built as the base tag, so as each platform builds it will clobber the previous push of that tag.

Reply:
Okay